java/code/src/com/redhat/rhn/domain/server/Server.java

```diff
@@ -2268,6 +2269,28 @@ public boolean isPayg() {
         return payg;
     }
 
+    /**
+     * @return true if this server is denied to be managed by SUSE Manager PAYG
+     */
+    public boolean isDeniedOnPayg() {
+        if (isPayg()) {
+            return false;
+        }
+        return !getInstalledProducts().stream()
+                .map(InstalledProduct::getSUSEProduct)
+                .filter(Objects::nonNull)
+                .allMatch(p -> {
+                    if (p.getFree()) {
+                        return true;
+                    }
+                    ChannelFamily cf = p.getChannelFamily();
+                    if (cf != null) {
+                        return cf.getLabel().startsWith("SMP") || cf.getLabel().equals("SLE-M-T");
+                    }
+                    return false;
+                });
```

Comment:
The initial `!`, plus the `allMatch`, plus returning what is instead allowed in the lambda might be confusing for someone not fully aware of the PAYG logic. A comment stating explicitly what we allow might be helpful here.

Reply:
Feel free to add a comment